Add extraction of protocol conformance descriptor info, as read from an object file image, to `TypeRefBuilder`.

artemcm opened the pull request:
This PR adds the ability to read protocol conformance descriptors from object files, in order to get at the type/protocol information. 

- Object file handling/reading is refactored into the `static_mirror` namespace: 
Code moved from `swift-reflection-dump.cpp` into `StaticMirror/ObjectFileContext.cpp` and `StaticMirror/ObjectFileContext.h`, without changes.
- `ObjectFile.h` has been taught about the `__swift5_proto` sections in relevant object file formats.
- `Metadata.h` now also provides type aliases for `External` versions of Conformance and Protocol descriptors, as opposed to `InProcess` ones.
- `TypeRefBuilder.h` now provides API to query and dump conformance information - `dumpConformanceSection`, which is templated on pointer size, since we may be reading this info from binaries built for 32-bit or 64-bit platforms.

artemcm committed:
Add Conformance section to `ReflectionInfo` read from object files

artemcm committed:
Factor out Object File handling from `swift-reflection-dump` to separate CMake target: StaticMirror

This target will then be used to have a broader set of tools for reading metadata from object files.

artemcm committed:
Add extraction protocol conformance descriptor extraction, as read from an object file image, to `TypeRefBuilder`.

mikeash commented on include/swift/ABI/Metadata.h:
```diff
@@ -79,6 +79,16 @@ struct RuntimeTarget<4> {
   using StoredPointerDifference = int32_t;
   static constexpr size_t PointerSize = 4;
+
+#if SWIFT_OBJC_INTEROP
```
You'll want to pull this out of `RuntimeTarget` and use `WithObjCInterop` or `NoObjCInterop` as appropriate instead. The idea is that the target's ObjC interop setting may not match what we're built with. See https://github.com/apple/swift/pull/40927/files for this recent change. /cc @adrian-prantl

adrian-prantl replied:
Yeah this is important. LLDB will instantiate various versions of this to support remote debugging where LLDB is built with SWIFT_OBJC_INTEROP but the remote target may not be.

artemcm replied:
Great catch - I'd already adjusted all the use-sites to use `WithObjCInterop` and `NoObjCInterop` and this was leftover from a previous iteration. Removed, thanks!

mikeash commented on include/swift/Reflection/ReflectionContext.h:
```diff
@@ -221,6 +221,22 @@ class ReflectionContext
+        // section name, to allow for matching of sections with a suffix
+        // like `_TEXT`, etc.
+        // "__swift5_proto" section name is a substring of "__swift5_protos",
+        // Ensure we don't return the latter when looking for the former.
```
Should the above really be looking for prefixes? I would expect section names to be exact. Segment names sometimes need prefix matching (e.g. `__DATA` becomes `__DATA_CONST`) but I don't think section names do. (I realize you didn't write the prefix check, but the right move here might be to make it an exact check rather than doing this additional check.)

mikeash replied:
The section name there is actually `__swift5_fieldmd`. It's just that MachO section names are 16 bytes, not arbitrary NUL-terminated strings. The `__TEXT` comes from what's immediately after. It's really common to mix these up, because almost all section names you encounter in the real world are less than 16 bytes and pad the end with zeroes.

I _think_ the right fix here is just to change `strlen(Name.data())` to `sizeof(S->sectname)`. That will compare up to the 16-byte limit.

artemcm replied:
Ah, that makes sense, thank you for the explanation! 
Using `strlen(Name.data()) ` does indeed achieve the expected behavior.

mikeash replied:
Excellent!

mikeash commented on include/swift/StaticMirror/ObjectFileContext.h:
```diff
@@ -0,0 +1,150 @@
+//===---------------- ObjectFileCOntext.h - Swift Compiler ---------------===//
```
Typo, `COntext`.

artemcm replied:
Fixed. Thanks!

mikeash commented on stdlib/public/SwiftRemoteMirror/SwiftRemoteMirror.cpp:
```diff
@@ -208,14 +208,16 @@ swift_reflection_addReflectionInfo(SwiftReflectionContextRef ContextRef,
     sectionFromInfo<const void *>(Info, Info.type_references),
-    sectionFromInfo<const void *>(Info, Info.reflection_strings)};
+    sectionFromInfo<const void *>(Info, Info.reflection_strings),
+    // TODO: Conformance section
```
`addReflectionInfo` is deprecated, so this can be left as `nullptr, 0` without the TODO. Clients should use `addImage` instead, which lets us decide which sections we need to locate.

mikeash reviewed (changes requested):
Some small comments, but this is looking very nice.

artemcm committed:
Add support for reading protocol descriptors from an indirect protocol descriptor target in an conformance descriptor.

Previously, the code assumed that such an indirect target will always point to an external symbol pointer, but it can also be an absolute pointer to an in-image protocol descriptor.

artemcm commented:
@swift-ci please test

artemcm committed:
Fix section name lookup in `findMachOSectionByName` to do comparison up-to the 16-byte section name limit, instead of the queried name length.

artemcm commented:
@swift-ci please test

slavapestov commented:
TypeRefBuilder was originally intended to work even without a remote process. Perhaps ReflectionContext would be a better place for something that reads sections?

artemcm committed:
Extract fully-qualified names for protocol conformances by traversing their parent contexts and extracting their names

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
> TypeRefBuilder was originally intended to work even without a remote process. Perhaps ReflectionContext would be a better place for something that reads sections?

I agree that `ReflectionContext` would probably be a better place for some of this functionality; but, moving over all of the section-reading code into there would involve refactoring a lot of other existing code in `TypeRefBuilder` that reads sections already. If it is alright with you, I'd prefer to tackle that refactor separately.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test macOS platform

artemcm committed:
[Static Mirror] Add mangled names to image-extracted protocol conformance infos.

Collect the mapping of unmangled-to-mangled names from the read out `Fields` section reflection infos and use that to identify the mangled name of a conforming type of a given protocol conformance.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test Windows platform

mikeash commented on include/swift/ABI/Metadata.h:
```diff
@@ -540,6 +540,14 @@ namespace {
+template<unsigned PointerSize>
+using ExternalTypeContextDescriptor
+#if SWIFT_OBJC_INTEROP
+= TargetTypeContextDescriptor<External<WithObjCInterop<RuntimeTarget<PointerSize>>>>;
```
This still isn't quite right, since we want the client to be able to specify whether the target has ObjC interop separately from how the client builds. I believe the right approach here would be to modify `ProtocolConformanceDescriptorReader` to be templated on ObjC interop as well as pointer size, and do the `WithObjCInterop`/`NoObjCInterop` logic there. @adrian-prantl Does that sound right?

artemcm replied:
Ah, makes perfect sense, thank you.
Updated with the type aliases templated on `ObjCInteropKind` to allow clients to decide on this separately from how this code is built.

artemcm committed:
[Static Mirror] Allow ObjCInteropKind to be specified by the client of `conformanceReader`, separately from how the client builds.

artemcm commented:
@swift-ci please test

nkcsgexi commented:
🎉
